pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,389 @@
+		endOffset += cb.config.ChunkBlockSize
+	}
+
+	blocks := (endOffset - startOffset) / cb.config.ChunkBlockSize
+
+	offsetKeys := make(map[int64]string, blocks)
+	keys := make([]string, 0, blocks)
+	lastBlockLength := 0
+
+	for o := startOffset; o < endOffset; o += cb.config.ChunkBlockSize {
+		end := o + cb.config.ChunkBlockSize
+		if end > int64(size) {
+			end = int64(size)
+		}
+
+		lastBlockLength = int(end - o)
+
+		k := cachingKeyObjectBlock(name, o, end)
+		keys = append(keys, k)
+		offsetKeys[o] = k
+	}
+
+	// Try to get all blocks from the cache.
+	hits, _ := cb.cache.Fetch(ctx, keys)
+
+	for _, b := range hits {
+		cb.cachedChunkBytes.Add(float64(len(b)))
+	}
+
+	if len(hits) < len(keys) {
+		if hits == nil {
+			hits = map[string][]byte{}
+		}
+
+		err := cb.fetchMissingChunkBlocks(ctx, name, startOffset, endOffset, offsetKeys, hits, lastBlockLength)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return newBlocksReader(cb.config.ChunkBlockSize, offsetKeys, hits, offset, length), nil
+}
+
+type rng struct {
+	start, end int64
+}
+
+// Fetches missing blocks and stores them into "hits" map.
+func (cb *CachingBucket) fetchMissingChunkBlocks(ctx context.Context, name string, startOffset, endOffset int64, cacheKeys map[int64]string, hits map[string][]byte, lastBlockLength int) error {
+	// Ordered list of missing sub-ranges.
+	var missing []rng
+
+	for off := startOffset; off < endOffset; off += cb.config.ChunkBlockSize {
+		if hits[cacheKeys[off]] == nil {
+			missing = append(missing, rng{start: off, end: off + cb.config.ChunkBlockSize})
+		}
+	}
+
+	missing = mergeRanges(missing, 0) // Merge adjacent ranges.
+	// Keep merging until we have only max number of ranges (= requests).
+	for limit := cb.config.ChunkBlockSize; cb.config.MaxChunksGetRangeRequests > 0 && len(missing) > cb.config.MaxChunksGetRangeRequests; limit = limit * 2 {
+		missing = mergeRanges(missing, limit)
+	}
+
+	var hitsMutex sync.Mutex
+
+	// Run parallel queries for each missing range. Fetched data is stored into 'hits' map, protected by hitsMutex.
+	eg, nctx := errgroup.WithContext(ctx)
+	for _, m := range missing {
+		m := m
+		eg.Go(func() error {
+			r, err := cb.bucket.GetRange(nctx, name, m.start, m.end-m.start)
+			if err != nil {
+				return errors.Wrapf(err, "fetching range [%d, %d]", m.start, m.end)
+			}
+			defer runutil.CloseWithLogOnErr(cb.logger, r, "fetching range [%d, %d]", m.start, m.end)
+
+			for off := m.start; off < m.end && nctx.Err() == nil; off += cb.config.ChunkBlockSize {
+				key := cacheKeys[off]
+				if key == "" {
+					return errors.Errorf("fetching range [%d, %d]: caching key for offset %d not found", m.start, m.end, off)
+				}
+
+				// We need a new buffer for each block, both for storing into hits, and also for caching.
+				blockData := make([]byte, cb.config.ChunkBlockSize)
+				n, err := io.ReadFull(r, blockData)
+				if err == io.ErrUnexpectedEOF && (off+cb.config.ChunkBlockSize) == endOffset && n == lastBlockLength {
```

Comment:
I think always considering `ErrUnexpectedEOF` safe for the last block is risky. Why don't we initialise `blockData` accurately for the last block and we never check for `err == io.ErrUnexpectedEOF`?

Reply:
> I think always considering `ErrUnexpectedEOF` safe for the last block is risky. 

There is one extra check -- whether `n` matches `lastBlockLength`, so there is no risk.

> Why don't we initialise `blockData` accurately for the last block and we never check for `err == io.ErrUnexpectedEOF`?

Yes, it will be nicer.